Add assertions for java.lang.reflect. Fix #3130

wjbakker opened the pull request:
This changes adds assertions for classes:
* Field
* Method
* Constructor

#### Check List:
* Fixes: #3130
* Unit tests : YES
* Javadoc with a code example (on API only) : YES
* PR meets the [contributing guidelines](https://github.com/assertj/assertj/blob/main/CONTRIBUTING.md)

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Improve javadoc.

wjbakker commented on assertj-core/src/main/java/org/assertj/core/error/ShouldHaveAnnotations.java:
```diff
@@ -31,13 +32,13 @@ public class ShouldHaveAnnotations extends BasicErrorMessageFactory {
    * @return the created {@code ErrorMessageFactory}.
    */
-  public static ErrorMessageFactory shouldHaveAnnotations(Class<?> actual,
+  public static ErrorMessageFactory shouldHaveAnnotations(AnnotatedElement actual,
```
The binary compatibility check fails on widening this static method from `Class<?>` to `AnnotatedElement`. Class implements this interface since Java 5. I'm not sure on requirements to keep backward compatible with pre Java 5 code.

scordio replied:
The main problem is not necessarily related to compatibility with Java < 5 but rather to the signature change of `shouldHaveAnnotations`, which can break code compiled with AssertJ <= 3.24.2.

Could you maybe overload it and deprecate the existing one?

wjbakker replied:
Done

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Fix binary compatibility for ShouldHaveAnnotations
Improve Javadoc in AbstractMethodAssert

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Formatting

scordio commented on assertj-core/src/main/java/org/assertj/core/api/Assertions.java:
```diff
@@ -306,6 +309,21 @@ public static MatcherAssert assertThat(Matcher actual) {
+   * @return the created assertion object.
+   * @since 3.25.0
+   */
+  public static AbstractConstructorAssert<?, Constructor<?>> assertThat(Constructor<?> actual) {
```
Yes, please always use `Abstract...Assert` return types to allow third-party libraries to provide their own extensions, see https://github.com/assertj/assertj/pull/2173#issuecomment-864465901.

We will clean up inconsistencies in AssertJ 4.

scordio replied:
Related: https://github.com/orgs/assertj/discussions/3055#discussioncomment-6035540

Commit:
Merge branch 'master' into 3130_java_lang_reflect_assertions

# Conflicts:
#	assertj-core/src/main/java/org/assertj/core/api/StandardSoftAssertionsProvider.java
#	assertj-core/src/main/java/org/assertj/core/internal/Classes.java
#	assertj-core/src/test/java/org/assertj/core/api/InstanceOfAssertFactoriesTest.java

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Change generic interface in Assertions to AbstractConstructorAssert<?, CLASS>

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Spotless formatting

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

License

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Change AssertionsForClassTypes to use Abstract...Assert

Commit:
Add assertions for java.lang.reflect. Fix #3130

This changes adds assertions for classes:
* Field
* Method
* Constructor

Add Assumptions
Add BDDAssertions
Add BDDAssumptions
Add unit tests
Update expected version to 3.27.0

wjbakker committed:
Merge branch 'assertj:3.x' into 3130_java_lang_reflect_assertions

wjbakker committed:
Merge branch 'assertj:3.x' into 3130_java_lang_reflect_assertions

wjbakker committed:
Merge branch 'assertj:3.x' into 3130_java_lang_reflect_assertions

wjbakker committed:
Merge branch 'assertj:3.x' into 3130_java_lang_reflect_assertions

wjbakker committed:
Merge branch 'assertj:3.x' into 3130_java_lang_reflect_assertions
